simulator/src/utils/prepare_instruction.rs

```diff
@@ -0,0 +1,519 @@
+        // another for the non-fungible ids. If there is more or less, then this function returns
+        // an error.
+        if tokens.len() != 2 {
+            return Err(ParseResourceSpecifierError::MoreThanOneAmountSpecified);
+        }
+
+        // Paring the resource address fully first to use it for the non-fungible id type ledger
+        // lookup
+        let resource_address_string = tokens[0];
+        let resource_address = bech32_decoder
+            .validate_and_decode_resource_address(resource_address_string)
+            .map_err(ParseResourceSpecifierError::InvalidResourceAddress)?;
+        let non_fungible_id_type = lookup_id_type(&resource_address)
+            .map_err(ParseResourceSpecifierError::LedgerLookupError)?;
+
+        // Parsing the non-fungible ids with the available id type
+        let non_fungible_ids = tokens[1]
+            .split(',')
+            .map(|s| s.trim())
+            .map(|s| NonFungibleId::try_from_simple_string(non_fungible_id_type, s))
+            .collect::<Result<BTreeSet<_>, _>>()
+            .map_err(ParseResourceSpecifierError::InvalidNonFungibleId)?;
+
+        Ok(ResourceSpecifier::Ids(non_fungible_ids, resource_address))
+    }
+}
+
+// ========
+// Errors
+// ========
+
+enum ParseResourceSpecifierError {
+    InvalidAmount(ParseDecimalError),
+    InvalidResourceAddress(AddressError),
+    InvalidNonFungibleId(ParseNonFungibleIdError),
+    LedgerLookupError(LedgerLookupError),
+    MoreThanOneAmountSpecified,
+}
+
+/// Represents an error when parsing arguments.
+#[derive(Debug, Clone)]
+pub enum BuildArgsError {
+    /// The argument is not provided.
+    MissingArgument(usize, Type),
+
+    /// The argument is of unsupported type.
+    UnsupportedType(usize, Type),
+
+    UnsupportedRootType(Type),
+
+    /// Failure when parsing an argument.
+    FailedToParse(usize, Type, String),
+
+    /// Failed to interpret this string as a resource specifier
+    InvalidResourceSpecifier(String),
+
+    /// Failed to perform a ledger lookup
+    LedgerLookupError(LedgerLookupError),
+}
+
+/// Represents an error when building a transaction.
+#[derive(Debug, Clone)]
+pub enum BuildCallWithAbiError {
+    /// The given blueprint function does not exist.
+    FunctionNotFound(String),
+
+    /// The given component method does not exist.
+    MethodNotFound(String),
+
+    /// The provided arguments do not match ABI.
+    FailedToBuildArgs(BuildArgsError),
+
+    /// Failed to export the ABI of a function.
+    FailedToExportFunctionAbi(PackageAddress, String, String),
+
+    /// Failed to export the ABI of a method.
+    FailedToExportMethodAbi(ComponentAddress, String),
+
+    /// Account is required but not provided.
+    AccountNotProvided,
+}
+
+impl From<BuildArgsError> for BuildCallWithAbiError {
+    fn from(error: BuildArgsError) -> Self {
+        Self::FailedToBuildArgs(error)
+    }
+}
```

Comment:
This should've been done a long time ago. Would appreciate it if we can add some unit tests for the supported types here.

Reply:
Sounds good, will do that!